### PR TITLE
Move address construction, encoding and decoding back to cardano-wallet-core

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#900 

# Overview

:information_source: I am sorry

- [x] I have adjusted `KeyToAddress` so that it'll use a `n :: Network` type parameter instead of a backend target .... and this ripples pretty much everywhere. 

- [x] I have moved `Network` from `Jormungandr.Environment` to `Primitive.AddressDerivation`.

- [x] I have moved `EncodeAddress` and `DecodeAddress` into `Primitive.AddressDiscovery`, and moved the internals decoding bits for Shelley and Byron into `Primitive.AddressDerivation.Sequential` and `Primitive.AddressDerivation.Random` respectively. 

- [x] I have moved `decodeLegacyAddress` from `Jormungandr.Binary` to `Primitive.AddressDerivation`

- [x] renamed `Cardano.Wallet.Primitive.AddressDerivation.Sequential` into `Cardano.Wallet.Primitive.AddressDerivation.Shelley`

- [x] renamed `Cardano.Wallet.Primitive.AddressDerivation.Random` into `Cardano.Wallet.Primitive.AddressDerivation.Byron`

- [x] renamed `Network` to `NetworkDiscriminant` to disambiguate from the networking layer

- [x] moved all the corresponding unit tests from Jörmungandr back to core too

# Comments

We currently use an opaque representation for `Address` which is very
convenient as it allows us to pass them around without really having to
worry about their internal structure.  Serialization and deserialization
is easy, and we can defer inspecting the internals only when really
needed.

Knowing the internal of addresses really matters only in a few places:

1. When performing random discovery, we need addresses to match the
Byron format (see
https://github.com/input-output-hk/cardano-wallet/wiki/About-Address-Format---Byron)
and carry their derivation path as a payload.

2. When performing sequential discovery, we need to be able to extract
the spending key (or a hash of it) from the address. This is possible
for both Byron and Jörmungandr addresses.

3. When generating change addresses, we need to know how to convert a
given spending key to an address.

4. When displaying a list of addresses from a list of known spending keys.

Our current codebase makes the distinction between `SeqKey` and `RndKey`
and implicitely assumes Byron addresses and random discovery for
`RndKey` and Jormungandr addresses with sequential discovery for
`SeqKey`. These assumptions are however false, since it is possible to
have sequential discovery also for Byron addresses (like Yoroi does).

In addition, the wallet engine is currently parameterized over a `s`,
`t` and `k :: Depth -> * -> *` type, which suppose then that a given
engine may only use one type of address per target. This again is a
false assumption since Jörmungandr comprehends both formats.

So, we have to review how we actually look at addresses and key
derivation such that:

- [x] Address format only depends on
  - The era / key type it's based upon
  - The network discriminant in use

- [x] Sequential discovery can be used with addresses or any format

### CODE REVIEW

**Comment by paweljakubas** [comment]
good so far! I totally buy into the idea that we need to treat addresses DIFFERENTLY and in such a way that parametrizing them by "era" is possible. And this allows seamlessly to service such combinations (also each mainnet/testnet): 
- Byron address + seq discovery (yoroi)
- Byron address + rand discovery 
- Jormungandr address + seq discovery

I am thinking now about them as concept that depends on network (mainnet/testnet), era (byron/shelley/...) and scheme (rnd/seq). And the picking of them should not be restricted. In terms of choice, ortogonality should be allowed

File: lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs [hunk 1]
```diff
@@ -368,7 +380,21 @@ instance MonadRandom ((->) (Passphrase "salt")) where
     getRandomBytes _ (Passphrase salt) = BA.convert salt
 
 {-------------------------------------------------------------------------------
-                          Backend-dependent functions
+                          Network Discrimination
+-------------------------------------------------------------------------------}
+
+-- | Available network options.
+data Network = Mainnet | Testnet
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs]: :100:  yes, this is the proper place for this

File: lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Sequential.hs [hunk 2]
```diff
@@ -350,6 +366,116 @@ deserializeXPrvSeq (k, h) = (,)
...
+    groupedAddressLength = 65
+    invalidFirstByte discriminant = TextDecodingError
+        $ "Invalid Address first byte: "
+        <> B8.unpack (BS.take 1 bytes)
+        <> " =/= "
+        <> B8.unpack (BS.pack [discriminant])
+        <> "."
+
+-- | Embed some constants into a network type.
+class KnownNetwork (n :: Network) where
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Sequential.hs]: I wonder if this is always the same irrespective of era (ie., the way addresses are discriminated) ... and hence this is the right place to put it

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Sequential.hs]: Note that I'll rename `AddressDerivation.Sequential` to `AddressDerivation.Shelley` in the end to better clarify the purpose of this module. You're right that this really depends on the era, and that's why the division here shouldn't be on the discovery scheme but on the era!

File: lib/core/cardano-wallet-core.cabal [hunk 3]
```diff
@@ -217,11 +219,12 @@ test-suite unit
...
       Cardano.Wallet.Network.BlockHeadersSpec
       Cardano.Wallet.NetworkSpec
-      Cardano.Wallet.Primitive.AddressDerivation.RandomSpec
-      Cardano.Wallet.Primitive.AddressDerivation.SequentialSpec
+      Cardano.Wallet.Primitive.AddressDerivation.ByronSpec
+      Cardano.Wallet.Primitive.AddressDerivation.ShelleySpec
       Cardano.Wallet.Primitive.AddressDerivationSpec
       Cardano.Wallet.Primitive.AddressDiscovery.RandomSpec
       Cardano.Wallet.Primitive.AddressDiscovery.SequentialSpec
+      Cardano.Wallet.Primitive.AddressDiscoverySpec
```
Comment (KtorZ) [inline lib/core/cardano-wallet-core.cabal:228]: This is where most of the `Cardano.Wallet.Jormungandr.CompatibilitySpec` specs have been moved.

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 4]
```diff
@@ -290,19 +293,19 @@ data Listen
...
+        , PaymentAddress n ByronKey
+        , PaymentAddress n ShelleyKey
         )
     => Warp.Settings
     -> Trace IO Text
     -> Socket
-    -> ApiLayer (RndState t) t RndKey
-    -> ApiLayer (SeqState t) t SeqKey
+    -> ApiLayer (RndState n) t ByronKey
+    -> ApiLayer (SeqState n) t ShelleyKey
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:308]: @paweljakubas now the relationship between era, keys and backend is made more obvious and here, we clearly see what we support. Yoroi is missing as being a combination of `SeqState` and `ByronKey`

**Comment by KtorZ** [comment]
bors try

**Review by paweljakubas — APPROVED** [review]
Very needed refactoring that will benefit the project in the future. Now we implicitly can get in starightforward way to any solution that have any network/era/scheme combination!. In the end when we have AddressPool revisited (#901 issue), then we will need to add API endpoints and handlers to support yoroi, haskell node, etc. if they comply with addresses standard, byron/shelley standard, and what mainnet/testnet means under the hood.  If not always we can introduce new address format that embraces that - `Cardano.Wallet.Primitive.AddressDerivation.SomethingNew` :100:

**Pull request merged**